### PR TITLE
Replace configmapprovider.Retrieved with returning config.Map and StopWatchFunc

### PR DESCRIPTION
**Pull request by bogdandrutu**
Fixes https://github.com/open-telemetry/opentelemetry-collector/issues/4665

Alternative considered: was to replace configmapprovider.Retrieved with a struct.

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: config/configmapprovider/provider.go [hunk 1]
```diff
@@ -66,6 +68,15 @@ type Provider interface {
 	Shutdown(ctx context.Context) error
 }
 
+// CloseFunc signals that the configuration for which it was used to retrieve values is
+// no longer in use and should close and release any watchers that it may have created.
+//
+// Should block until all resources are closed, and guarantee that `onChange` is not
+// going to be called after it returns except when `ctx` is cancelled.
+//
+// TODO: Find a good name for this func.
```
Comment (bogdandrutu) [inline config/configmapprovider/provider.go:77]: Any ideas?

**Review by bogdandrutu — COMMENTED** [review]
Will fix the changelog when I get an approver :)

File: config/configmapprovider/provider.go (hunk 1 above)
Comment (Aneurysm9) [reply config/configmapprovider/provider.go:77]: This feels like a reasonable name.  Externally, it'll be referenced as `configmapprovider.CloseFunc`, which is a bit of a mouthful but has all the relevant info.

If this doesn't feel right, what about a `Close` method on the `Provider` interface?

**Review by Aneurysm9 — COMMENTED** [review]
Generally looks good.  I think it could maybe be simplified further by making the `CloseFunc` a `Close` method on the `Provider` interface.  Would be interested in hearing if you had reasons for not doing that.

**Comment by bogdandrutu** [comment]
@Aneurysm9 close right now is designed to stop/close (maybe stop watching func is a better name because of this) watching for the specified "location" in the retrieve, not for all the locations. So we can move it to provider if we change the behavior as well, but that has less flexibility for the future.

**Comment by Aneurysm9** [comment]
> @Aneurysm9 close right now is designed to stop/close (maybe stop watching func is a better name because of this) watching for the specified "location" in the retrieve, not for all the locations. So we can move it to provider if we change the behavior as well, but that has less flexibility for the future.

Ok, I think I've got it now.  I think for some reason I was assuming earlier binding of the location.  This does seem like the right API given that the stop/close is scoped to the retrieved location and the provider is not.

> Please read the API description (what is now merged, and what is here) and let me know if that is not clear about the behavior I mentioned, because we need to improve it otherwise.

The description of "typical usage" of the `Provider` interface does need to be updated as it no longer reflects the expected usage pattern.  But I don't think that my misunderstanding is a fault of the description.

File: config/configmapprovider/provider.go (hunk 1 above)
Comment (bogdandrutu) [reply config/configmapprovider/provider.go]: I tried `StopWatchFunc`, if not better I can revert.

**Comment by bogdandrutu** [comment]
> The description of "typical usage" of the Provider interface does need to be updated as it no longer reflects the expected usage pattern. 

Done.

File: config/configmapprovider/provider.go (hunk 1 above)
Comment (Aneurysm9) [reply config/configmapprovider/provider.go]: That makes me think of a timer/stopwatch.  Maybe `EndWatchFunc`?

File: config/configmapprovider/provider.go [hunk 2]
```diff
@@ -24,15 +26,13 @@ import (
 //
 // The typical usage is the following:
 //
-//		r := mapProvider.Retrieve("file:/path/to/config")
-//		r.Get()
+//		cfg, stopWatch, err := mapProvider.Retrieve("file:/path/to/config")
```
Comment (Aneurysm9) [inline config/configmapprovider/provider.go:29]:
```suggestion
//		cfg, stopWatch, err := mapProvider.Retrieve(ctx, "file:/path/to/config", onChange)
```

File: config/configmapprovider/provider.go [hunk 3]
```diff
@@ -24,15 +26,13 @@ import (
...
 //		// wait for onChange() to be called.
-//		r.Close()
-//		r = mapProvider.Retrieve("file:/path/to/config")
-//		r.Get()
+//		if stopWatch != nil { stopWatch() }
+//		cfg, stopWatch, err = mapProvider.Retrieve("file:/path/to/config")
```
Comment (Aneurysm9) [inline config/configmapprovider/provider.go:32]:
```suggestion
//		cfg, stopWatch, err = mapProvider.Retrieve(ctx, "file:/path/to/config", onChange)
```

File: config/configmapprovider/provider.go (hunk 1 above)
Comment (bogdandrutu) [reply config/configmapprovider/provider.go]: Happy to try that, between `EndWatchFunc` or `CloseWatchFunc` or `DoneFunc` (does not seem to end, but saw this pattern)?

Comment (Aneurysm9) [reply config/configmapprovider/provider.go]: `DoneFunc` also seems good.  I think I would be happy with either that or `EndWatchFunc`.  Would like to know what others think.

Comment (bogdandrutu) [reply config/configmapprovider/provider.go]: FYI: Usually (for unknown reasons) you don't get feedback unless you tag people :) @tigrannajaryan @codeboten @jpkrohling @dmitryax

File: config/configmapprovider/provider.go (hunk 2 above)
Comment (bogdandrutu) [reply config/configmapprovider/provider.go:29]: Done.

File: config/configmapprovider/provider.go (hunk 3 above)
Comment (bogdandrutu) [reply config/configmapprovider/provider.go:32]: Done.

File: config/configmapprovider/provider.go (hunk 1 above)
Comment (bogdandrutu) [reply config/configmapprovider/provider.go]: Tried `EndWatchFunc` looks ok...

File: CHANGELOG.md [hunk 4]
```diff
@@ -19,6 +19,7 @@
 - Deprecate `configtelemetry.Level.Set()` (#4700)
 - Remove support to some arches and platforms from `ocb` (opentelemetry-collector-builder) (#4710)
 - Remove deprecated legacy path ("v1/trace") support for otlp http receiver (#4720)
+- Replace configmapprovider.Retrieved with returning config.Map and StopWatchFunc (#4721)
```
Comment (Aneurysm9) [inline CHANGELOG.md:22]:
```suggestion
- Replace configmapprovider.Retrieved with returning config.Map and EndWatchFunc (#4721)
```

File: config/configmapprovider/provider.go [hunk 5]
```diff
@@ -24,15 +26,13 @@ import (
 //
 // The typical usage is the following:
 //
-//		r := mapProvider.Retrieve("file:/path/to/config")
-//		r.Get()
+//		cfg, stopWatch, err := mapProvider.Retrieve(ctx, "file:/path/to/config", onChange)
```
Comment (Aneurysm9) [inline config/configmapprovider/provider.go:29]:
```suggestion
//		cfg, endWatch, err := mapProvider.Retrieve(ctx, "file:/path/to/config", onChange)
```

GitHub won't let me make a single suggestion here, but apply to the rest of this docblock.

**Review by Aneurysm9 — APPROVED** [review]

File: config/configmapprovider/file.go [hunk 6]
```diff
@@ -47,25 +47,23 @@ func NewFile() Provider {
 	return &fileMapProvider{}
 }
 
-func (fmp *fileMapProvider) Retrieve(_ context.Context, location string, _ WatcherFunc) (Retrieved, error) {
+func (fmp *fileMapProvider) Retrieve(_ context.Context, location string, _ WatcherFunc) (*config.Map, EndWatchFunc, error) {
```
Comment (jpkrohling) [inline config/configmapprovider/file.go:50]: Given this is an API, would it be better to return a specific type containing the map and the func? A func that returns 3 values leaves a bad aftertaste, especially for a pre-v1 API, in my opinion.

Comment (bogdandrutu) [reply config/configmapprovider/file.go:50]: So you think the alternative consideration is better. I think having a simple struct with 2 public members is as good as this and definitely better than an interface (as it was before).

Comment (jpkrohling) [reply config/configmapprovider/file.go:50]: My reasoning is that it's easy to add a new field to the return object type if needed, without breaking current callers, while if you add an extra value to the return values, you break them. For some reason, I think it's easier to start needing 4 values when you have 3 than needing 3 when you have 2 (one of them being `error`).

Comment (bogdandrutu) [reply config/configmapprovider/file.go:50]: Sounds reasonable to me, will either change this or maybe do another one to look at that option as well.

Comment (bogdandrutu) [reply config/configmapprovider/file.go:50]: https://github.com/open-telemetry/opentelemetry-collector/pull/4789 here is the alternative, let me know.

**Comment by bogdandrutu** [comment]
Based on the comments in the alternative PR https://github.com/open-telemetry/opentelemetry-collector/pull/4789, closing this.

**Pull request closed without merging**